src/main/java/de/metas/ui/web/order/sales/purchasePlanning/view/PurchaseRowFactory.java

```diff
@@ -52,177 +56,146 @@
-					vendorProductInfo.getProductNo(),
+		if (vendorProductInfo != null) {
+			productId = vendorProductInfo.getProductId();
+			product = createProductLookupValue(vendorProductInfo.getProductId(), vendorProductInfo.getProductNo(),
 					vendorProductInfo.getProductName());
-		}
-		else
-		{
-			product = createProductLookupValue(purchaseCandidate.getProductId());
+		} else {
+			productId = purchaseCandidate.getProductId();
+			product = createProductLookupValue(productId);
 		}
 		final String uom = createUOMLookupValueForProductId(product.getKeyAsInt());
 
-		final int processedPurchaseCandidateId = purchaseCandidate.isProcessed() ? purchaseCandidate.getPurchaseCandidateId() : 0;
+		final int processedPurchaseCandidateId = purchaseCandidate.isProcessed()
+				? purchaseCandidate.getPurchaseCandidateId()
+				: 0;
 
 		return PurchaseRow.builder()
-				.rowId(PurchaseRowId.lineId(
-						purchaseCandidate.getSalesOrderLineId(),
-						bpartnerId,
+				.rowId(PurchaseRowId.lineId(purchaseCandidate.getSalesOrderLineId(), bpartnerId,
 						processedPurchaseCandidateId))
-				.salesOrderId(purchaseCandidate.getSalesOrderId())
-				.rowType(PurchaseRowType.LINE)
-				.product(product)
-				.uomOrAvailablility(uom)
-				.qtyToPurchase(purchaseCandidate.getQtyToPurchase())
-				.purchasedQty(purchaseCandidate.getPurchasedQty())
-				.datePromised(datePromised)
-				.vendorBPartner(vendorBPartner)
-				.purchaseCandidateId(purchaseCandidate.getPurchaseCandidateId())
-				.orgId(purchaseCandidate.getOrgId())
-				.warehouseId(purchaseCandidate.getWarehouseId())
-				.readonly(purchaseCandidate.isProcessedOrLocked())
-				.build();
+				.salesOrderId(purchaseCandidate.getSalesOrderId()).rowType(PurchaseRowType.LINE).product(product)
+				.uomOrAvailablility(uom).qtyToPurchase(purchaseCandidate.getQtyToPurchase())
+				.purchasedQty(purchaseCandidate.getPurchasedQty()).datePromised(datePromised)
+				.vendorBPartner(vendorBPartner).purchaseCandidateId(purchaseCandidate.getPurchaseCandidateId())
+				.orgId(purchaseCandidate.getOrgId()).warehouseId(purchaseCandidate.getWarehouseId())
+				.readonly(purchaseCandidate.isProcessedOrLocked()).build();
 	}
 
-	public PurchaseRow createGroupRow(final I_C_OrderLine salesOrderLine, final List<PurchaseRow> rows)
-	{
+	public PurchaseRow createGroupRow(final I_C_OrderLine salesOrderLine, final List<PurchaseRow> rows) {
 		final JSONLookupValue product = createProductLookupValue(salesOrderLine.getM_Product_ID());
 		final BigDecimal qtyToDeliver = salesOrderLine.getQtyOrdered().subtract(salesOrderLine.getQtyDelivered());
 		final String uom = createUOMLookupValueForProductId(product.getKeyAsInt());
 
+		final AvailableToPromiseQueryBuilder atpQueryBuilder = AvailableToPromiseQuery.builder();
+
+		atpQueryBuilder.productId(salesOrderLine.getM_Product_ID());
+		atpQueryBuilder.date(salesOrderLine.getDatePromised());
+
+		final BigDecimal qtyAvailable = new AvailableToPromiseRepository()
```

Comment:
AvailableToPromiseRepository is a spring boot service, so please let spring maintain its life cycle.
E.g. take a look at e.g. SalesOrder2PurchaseViewFactory which has spring inject *this* (PurchaseRowFactory) into itself